src/subprocess-posix.cc

```diff
@@ -75,6 +79,14 @@ bool Subprocess::Start(SubprocessSet* set, const string& command) {
   if (err != 0)
     Fatal("posix_spawnattr_init: %s", strerror(err));
 
+  if(priority != INT_MIN) {
+    struct sched_param sp{};
+    sp.sched_priority = priority;
+    err = posix_spawnattr_setschedparam(&attr, &sp);
+    if(err != 0)
+      Fatal("setpriority: %s", strerror(errno));
+  }
```

Comment:
This is correct, but I believe this will be ignored if you do not add `POSIX_SPAWN_SETSCHEDPARAM` to the `flags` value below.

Reply:
In the latest commit, I set the POSIX_SPAWN_SETSCHEDPARAM flag, but when setting any priority other than 0 I get `fatal: posix_spawn: Invalid argument`

Reply:
Ah. Maybe initializing the value by calling pthread_getschedparam() instead of zero-ing the struct will solve it?

Reply:
Sadly, this doesn't fix it 😢 I've now also looked into the man pages for some of these functions but didn't find anything obviously wrong with my implementation...

Reply:
I do not see this latest commit you are referencing. Can you upload a new version of this commit that sets the flag properly and initialized the params structure with pthread_getschedparam() so we can have a look at it?

Otherwise, the GLibc sources for posix_spawn() do not seem to return EINVAL explicitly, which suggest that one system call is returning this value. 

This maybe the call to `__sched_setparam()` in that source file, which according to the manpage should return EINVAL only if "the argument param does not make sense              for the current scheduling policy" !?

[1] https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/unix/sysv/linux/spawni.c;h=e8ed2babb9b1396903466f631da0571504896aa5;hb=143ef68b2aded7c794956beddad495af8c7d3251

[2] https://man7.org/linux/man-pages/man2/sched_setparam.2.html

Reply:
Yeah, sorry, forgot to push the change. I've now debugged the execution with rr and stepped back to where it returned `EINVAL`. It's coming from here: __pthread_setcancelstate https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/pthread_setcancelstate.c;h=18fb42a5c4538678358f744b5f356eca308d8401;hb=HEAD#l29
Also for me the sched_param struct has only the sched_priority field, so using posix_spawnattr_getschedparam first doesn't change anything.

Reply:
Weird, as far as I see, `pthread_setcancelstate()` does not modify errno, and the sources I linked to seem to call `__pthread_setcancelstate()` and never look at the returned value (and the argument seems correct, and would not trigger this EINVAL). Are you sure this is where the return error value is coming?

That doesn't make much sense, this there is no connection to scheduling parameters here at all, so this is extra puzzling :-/

Reply:
I have downloaded your PR and fixed it a little (i.e. you should call `pthread_getschedparam()` not `posix_spawnattr_setschedparam()` to get the current thread's scheduler parameters, then the SubprocessSet.ConsoleTest unit-test needs to have its SubprocessSet::Add() call adjusted for the new `priority` parameter).

I confirm that using a non-0 priority results in an EINVAL error when calling the `sched_setparam()` system call, as seen using `strace -f build-cmake/ninja_test --gtest_filter=SubprocessSet.ConsoleTest` which ends up with:

```
[pid 2490739] rt_sigaction(SIGRT_31, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7fcc5f97b510}, NULL, 8) = 0
[pid 2490739] rt_sigaction(SIGRT_32, NULL, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=0}, 8) = 0
[pid 2490739] rt_sigaction(SIGRT_32, {sa_handler=SIG_DFL, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7fcc5f97b510}, NULL, 8) = 0
[pid 2490739] sched_setparam(0, [1])    = -1 EINVAL (Invalid argument)
[pid 2490739] exit_group(127)           = ?
[pid 2490738] <... clone3 resumed>)     = 2490739
[pid 2490739] +++ exited with 127 +++
wait4(2490739, NULL, 0, NULL)           = 2490739
munmap(0x7fcc5ff15000, 36864)           = 0
rt_sigprocmask(SIG_SETMASK, [HUP INT TERM], NULL, 8) = 0
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=2490739, si_uid=28587, si_status=127, si_utime=0, si_stime=0} ---
write(2, "ninja: fatal: ", 14ninja: fatal: )          = 14
write(2, "posix_spawn: Invalid argument", 29posix_spawn: Invalid argument) = 29
```
So it looks like the kernel is complaining when `priority == 1`.

Reply:
Oh, the manpage for sched_get_priority_min() says:

```
Linux allows the static priority value range 1 to 99 for SCHED_FIFO and SCHED_RR and the priority 0 for SCHED_OTHER and SCHED_BATCH. Scheduling priority ranges for the various policies are not alterable.
```

So it looks like 0 is the only valid value for SCHED_OTHER or SCHED_BATCH. I think the solution is simply to restrict the priority passed to the spawn call to the valid range for the current policy in the code. I.e. something like:

```
    int policy = -1;
    err = pthread_getschedparam(pthread_self(), &policy, &sp);
    if (err != 0)
      Fatal("pthread_getschedparam: %s", strerror(err));

   int min_priority = sched_get_priority_min(policy);
   int max_priority = sched_get_priority_max(policy);
   if (priority < min_priority)
       priority = min_priority;
   else if (priority > max_priority)
       priority = max_priority;
```

With this, all tests pass, though I assume the priority value is essentially ignored as it is always clamped to 0 :-/

Reply:
Yeah, you are right __pthread_setcancelstate returns EINVAL, but isn't the actual cause for the error of posix_spawn. This seems to happen in the clone3 syscall.

Reply:
But if we can't set the priority to something other than 0, why can I set the priority to other values with the nice tool (GNU coreutils)?

Reply:
So in conclusion, there is a drastic difference between `setpriority()` and `sched_setparam()`, and I'm not sure to know why. This Linux source file seems to indicate that the kernel scheduler differentiates between the two:

https://github.com/torvalds/linux/blob/dccb07f2914cdab2ac3a5b6c98406f765acab803/kernel/sched/core.c#L7935

So I guess we should forget about `spawn_attr_set_schedparam()`, sorry for giving you this idea. I recommend you use `getpriority() / setpriority()` before the spawn() to save then change the niceness of the current (Ninja process) before the fork (so it is inherited by the child one), then call `setpriority()` just after that to restore the original niceness. And add one big comment to explain the reason!

What do you think?

Reply:
I've implemented that, but somehow restoring the previous priority after the posix_spawn results in a "Permission denied" error :cry: